### DISCUSSION TITLE
SellingApiBadRequestException

### DISCUSSION THREAD

**Discussion by smartguyscode** [saleweaver/python-amazon-sp-api#282] [Q&A]
Hi,

Please accept my apologies if this is a silly question, but I am a bit stuck and would really appreciate any help the community can provide.

I am trying to use this package to create multi channel fulfilment orders.  I get the following error when I try to make my call:

SellingApiBadRequestException: [{'code': 'InvalidInput', 'message': 'missing value for timestamp type', 'details': ''}]

Note that I have been able to successfully authenticate, and running the example provided works great:(orders(credentials=credentials).get_orders(CreatedAfter=(datetime.utcnow() - timedelta(days=7)).isoformat()))

Detail:
credentials=dict(
        refresh_token='xxx',
        lwa_app_id='xxx',
        lwa_client_secret='xxx',
        aws_secret_key='xxx',
        aws_access_key='xxx',
        role_arn='xxx',
    )

testpayload=json.dumps({
                        "marketplaceId": "A39IBJ37TRP1C6",
                        "sellerFulfillmentOrderId": "FBATestOrder-11",
                        "displayableOrderId": "TestOrder-FBAOutbound",
                        "displayableOrderDate": "2020-01-09T19:46:45.809Z",
                        "displayableOrderComment": "TestOrder",
                        "shippingSpeedCategory": "Standard",
                        "destinationAddress": {
                          "name": "Amazon",
                          "addressLine1": "20 George Street",
                          "addressLine2": "Suite 123",
                          "addressLine3": "Lane1",
                          "stateOrRegion": "NSW",
                          "countryCode": "AU",
                          "city": "Sydney",
                          "postalCode": "2000"
                        },
                        "fulfillmentAction": "Hold",
                        "notificationEmails": [
                          "test1%40amazon.com",
                          "test2%40amazon.com"
                        ],
                        "items": [
                          {
                            "sellerSku": "SG-SH-WF-RE",
                            "sellerFulfillmentOrderItemId": "OrderItemID1",
                            "quantity": 3
                          }
                        ]
                })

FulfillmentOutbound(credentials=credentials).create_fulfillment_order(body=testpayload)

**Comment by heysander** [comment]
Did you solve this one? Getting the same error now... Google is not very helpful, but digging through the code and hope to solve it soon :-)

**Reply by saleweaver** [reply]
I haven't worked with the endpoint before, but you usually never need to `json.dumps` anything

**Reply by heysander** [reply]
Thank you for your answer.
We don't use json.dumps. 
This is the full exception output:

```
raised unexpected: SellingApiBadRequestException([{'code': 'InvalidInput', 'message': 'missing value for timestamp type', 'details': ''}])
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/celery/app/trace.py", line 412, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
  File "/Users/***.py", line 491, in ***
    req = r.create_fulfillment_order(body=payload)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sp_api/base/helpers.py", line 19, in wrapper
    return function(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sp_api/api/fulfillment_outbound/fulfillment_outbound.py", line 331, in create_fulfillment_order
    return self._request(kwargs.pop('path'),  data=kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sp_api/base/client.py", line 129, in _request
    return self._check_response(res)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sp_api/base/client.py", line 144, in _check_response
    raise exception(error, headers=res.headers)
sp_api.base.exceptions.SellingApiBadRequestException: [{'code': 'InvalidInput', 'message': 'missing value for timestamp type', 'details': ''}]
```

**Reply by heysander** [reply]
Still figuring this error out. All three timestamps in the body ( displayableOrderDate, deliveryWindow{ startDate, endDate } ) are formatted like this 2022-05-14T16:25:24Z following the documentation this should be the right way, but the error says it is missing a value in the timestamp type... Cannot find any elaboration on this error message online.

**Reply by abrihter** [reply]
@sandersuitup
Do we have any new info regarding this one?

**Comment by tbrains** [comment]
@abrihter @sandersuitup @smartguyscode pass payload dict as keyword arguments

`FulfillmentOutbound(credentials=credentials).create_fulfillment_order(**testpayload)`

**Reply by heysander** [reply]
Hey @tbrains thanks for the suggestion, I quickly checked and it solved the datetime error! 
It leaves us with a new error:
[{'code': 'InvalidInput', 'message': 'The following required parameters are missing: [PostalCode, City]', 'details': ''}]
I will do some research on this one and if I find a solution, i'll share in this thread.

**Reply by tbrains** [reply]
@sandersuitup ,

Check if you are passing the parameters (PostalCode, City) in **destinationAddress**. I have attached a sample payload below.

```
{
    "marketplaceId": "ATVPDKIKX0DER",
    "sellerFulfillmentOrderId": "111222333",
    "displayableOrderId": "T-111222333",
    "displayableOrderDate": "2022-08-24T10:51:53Z",
    "displayableOrderComment": "TestOrder",
    "shippingSpeedCategory": "Standard",
    "deliveryWindow": {
        "startDate": "2022-08-24T10:51:53Z",
        "endDate": "2022-08-30T10:51:53Z"
    },
    "destinationAddress": {
        "name": "Amazon",
        "addressLine1": "ADDRESS 1",
        "addressLine2": "",
        "addressLine3": "",
        "stateOrRegion": "CA",
        "countryCode": "US",
        "city": "SACRAMENTO",
        "postalCode": "94203"
    },
    "fulfillmentAction": "Hold",
    "items": [
        {
            "sellerSku": "SG-SH-WF-RE",
            "sellerFulfillmentOrderItemId": "OrderItemID1",
            "quantity": 1,
            "perUnitDeclaredValue": {
                "currencyCode": "USD",
                "value": 50
            }
        }
    ]
}
```

**Reply by heysander** [reply]
This works great! We had "address" instead of "destinationAddress"
Thanks a lot!

**Reply by abrihter** [reply]
@tbrains 
Thank you :)

**Reply by nileshchauhan117** [reply]
I'm into similar problem now:

Request::: {"sellerFulfillmentOrderId":"PEER1681558668","displayableOrderId":"PEER1681558668","displayableOrderDate":"2023-04-15T11:37:48Z","displayableOrderComment":"Thank you.","shippingSpeedCategory":"Standard","destinationAddress":{"name":"john","addressLine1":"Some_Address","city":"Greensboro","stateOrRegion":"NC","countryCode":"US","postalCode":"27400"},"items":[{"sellerSku":"8XXX002","quantity":1,"sellerFulfillmentOrderItemId":"BXXXNM5XBY"}]}

Response::: {"errors":[{"code": "InvalidInput","message": "missing value for timestamp type","details": ""}]}

@sandersuitup  @abrihter @tbrains Can you spot the error here?
